### ISSUE TITLE
The "base" keyword requiring subclasses to also use "base/sealed/final" is inconvenient for testing purposes

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/language#3422]
Currently when a class is marked as "base", subclasses need to also be marked as such (or be sealed/final).
While the logic makes sense, it puts a significant burden on the syntax.

As a package author, I have a few use-cases where an abstract class would benefit from being "base" to prevent mocking mistakes.

TL;DR, instead of:

```dart
// from my package
abstract class Base {}

// in user code:
class Example extends Base {}

// Incorrect test code:
class ExampleMock with Mock implements Example {}
```

They should do:

```dart
class ExampleMock extends Base with Mock implements Example {}
```

Such that this mocks methods introduced by `Example` but not the one from the `Base` class.

The issue is:
- Only a minority of users need to mock such a class.
  But if `Base` was marked with "base", all users would have to make their subclass of `Base` with `base/final/sealed` – even if not necessary.
  This adds a cost to the majority of for a minority of cases
- There is a clear solution, and the mistake only happens during testing.
  As such, not using the `base` keyword doesn't hinder production

So as a package author, I'm in a weird conundrum. On one side, that keyword would help users spot their mistake more naturally due to writing `class ExampleMock with Mock implements Example {}` now being a compilation error.  
On the flip side, the gain is too niche to justify asking everyone to modify all their subclasses.

I'm not sure what the solution would be, as I understand subclasses are required to use a keyword. We probably want to avoid cases where a class _looks_ like it could be implemented, but can't due to a transitive superclass being "base".

**Comment by eernstg** [comment]
See https://github.com/dart-lang/language/issues/3111 for a discussion about an idea that might fit for you: We could allow a `base` class to be used in an `implements` clause if the given class declaration is a mock class.

There could be lots of ways to determine that a class is a mock class, but the approach taken there is that it is a mock class if and only if it has a non-trivial `noSuchMethod`.

However, that proposal didn't get much support, as described [here](https://github.com/dart-lang/language/issues/3111#issuecomment-1569706685).

**Comment by rrousselGit** [comment]
> We could allow a base class to be used in an implements clause if the given class declaration is a mock class.

That is the very opposite of what I want.

I want to use "base" only to prevent said class from being implemented by mocks

**Comment by eernstg** [comment]
Oh, I thought you mentioned two different versions of `ExampleMock` ('incorrect' and 'They should do') because you wanted to write a mock class, and you wanted it to be somehow less verbose.

> I want to use "base" only to prevent said class from being implemented by mocks

But if you want to use `base` _to prevent said class from being implemented by mocks_ then you're basically done: It is at least not easy to declare a mock of a `base` class. Discussion about this very topic [here](https://github.com/dart-lang/language/issues/3106).

If you want to use `base` to prevent said class from being implemented by mocks, _and that keyword should not have any other effects_ then it gets difficult: We don't have any firm definition of what it takes for a class to be a mock class, so presumably you can never prevent mock classes unless you prevent all subtypes whatsoever (e.g., by making the class `final` and not providing any subtypes in the same library).

**Comment by rrousselGit** [comment]
My issue isn't about the behaviour of using `base` with mocks.
The keyword behaves like I want it to.

The problem is about how subclasses now have to also specify `base`/`final`/`sealed`. I understand why, but that's a pain.

I've never had a single user make the mistake of _implementing_ my class _outside of tests_.
The mistake only happens inside tests due to mocking for my class.

So I'm torn between "adding `base` to the class to prevent the mock mistake people do" and "not adding that keyword to ease-up the syntax for every other usages".   
The testing issue is quite minor overall. But asking everyone to specify an extra keyword on their class is a non-negligible cost.

**Comment by eernstg** [comment]
By the way, if you _don't_ require all subtypes of a `base` declaration to be `base` then that keyword wouldn't provide any guarantees at all, it just becomes a comment. With the current design, with a `base` class or mixin `B` in a library _L_, every subtype of `B` is a subclass of a declaration in _L_ which is a subtype of `B`, and this makes it an a local (and rather easy) check that every subtype of `B` will indeed have an implementation of every private member of `B`. That again makes it safe to use an expression of type `B` as a receiver of invocations of those private members, which is significantly different from just having a keyword spelled `base` that means nothing.

**Comment by eernstg** [comment]
> But asking everyone to specify an extra keyword on their class is a non-negligible cost.

OK, that's a point of view that we haven't discussed much in the language team. I do agree that it might be considered a nuisance to write `base` on a large number of classes, but the other side of that coin is that it could be considered a significant and inconvenient surprise if a class/mixin/etc. could be `base`, and it wasn't shown locally on the declaration itself: Being `base` is a significant commitment for a class ("this class/mixin/etc. is sort-of owned by the library/libraries that declare the most general superinterfaces that are `base`, and you will never be able to drop the connection").

**Comment by rrousselGit** [comment]
I agree on that downside of that previous proposal. Hence why I wanted to avoid making that request directly and rather talk about the issue without a proposal

Still, in my case, since the issue only happens during tests (again, I haven't seen anyone make the mistake outside of mocks), the cost of `base` makes it very difficult for me to justify using the keyword – even though using it makes total sense.

**Comment by eernstg** [comment]
> Hence why I wanted to avoid making that request directly and rather talk about the issue without a proposal

Makes sense!

**Comment by leafpetersen** [comment]
> The problem is about how subclasses now have to also specify `base`/`final`/`sealed`. I understand why, but that's a pain.
> My class is very often subclassed by users. So using it would require lots of code to be updated to add an extra keyword on all subclasses.

We discussed this very extensively during the design phase, and the eventual consensus we came to was that we felt it was valuable for readers of code to be able to see precisely what the capabilities of a class were by looking at its declaration.  It feels detrimental to readability if you have to walk up the entire super-interface hierarchy to know whether or not a class has an interface.  cc @munificent may have more thoughts.

**Comment by lrhn** [comment]
The forced documentation was the primary reason for requiring `base` on types that have to be `base`. That, and failing early.
If someone adds `base` to a superclass, then every subclass also has to be `base`. If they aren't, it's immediately an error.
If we allowed subclasses to be implicitly `base`, without writing it, then subclasses in another package could live in happy ignorance that they've been turned `base`, until the owner tries to actually implement one.
And you could argue that that's great, as long as it works, it works.

Generally we try to aim for "explicit over implicit", because it lets you address issues earlier. Because when it stops working, you often wish you could have done something about it much earlier.
